flink-table-store-common/src/main/java/org/apache/flink/table/store/data/LazyGenericRow.java

```diff
@@ -0,0 +1,171 @@
+ */
+
+package org.apache.flink.table.store.data;
+
+import org.apache.flink.table.store.types.RowKind;
+
+import java.util.function.Supplier;
+
+import static org.apache.flink.table.store.utils.Preconditions.checkNotNull;
+
+/** A {@link InternalRow} which lazy init fields. */
+public class LazyGenericRow implements InternalRow {
+
+    /** The kind of change that a row describes in a changelog. */
+    private RowKind kind;
+
+    /** The array to store field value suppliers. */
+    private final Supplier<Object>[] suppliers;
+
+    /** The array to store the actual internal format values. */
+    private final Object[] fields;
+
+    /** The array to store field initialize status. */
+    private final boolean[] initialized;
+
+    public LazyGenericRow(Supplier<Object>[] suppliers) {
+        this(RowKind.INSERT, suppliers);
+    }
+
+    public LazyGenericRow(RowKind rowKind, Supplier<Object>[] suppliers) {
+        this.kind = rowKind;
+        this.suppliers = suppliers;
+        this.fields = new Object[suppliers.length];
+        this.initialized = new boolean[suppliers.length];
+    }
+
+    /**
+     * Returns the field value at the given position.
+     *
+     * <p>Note: The returned value is in internal data structure. See {@link InternalRow} for more
+     * information about internal data structures.
+     *
+     * <p>The returned field value can be null for representing nullability.
+     */
+    public Object getField(int pos) {
+        if (!initialized[pos]) {
+            fields[pos] = suppliers[pos].get();
+            initialized[pos] = true;
+        }
+        return fields[pos];
+    }
+
+    @Override
+    public int getFieldCount() {
+        return fields.length;
+    }
+
+    @Override
+    public RowKind getRowKind() {
+        return kind;
+    }
+
+    @Override
+    public void setRowKind(RowKind kind) {
+        checkNotNull(kind);
+        this.kind = kind;
+    }
+
+    @Override
+    public boolean isNullAt(int pos) {
+        return getField(pos) == null;
+    }
+
+    @Override
+    public boolean getBoolean(int pos) {
+        return (boolean) getField(pos);
+    }
+
+    @Override
+    public byte getByte(int pos) {
+        return (byte) getField(pos);
+    }
+
+    @Override
+    public short getShort(int pos) {
+        return (short) getField(pos);
+    }
+
+    @Override
+    public int getInt(int pos) {
+        return (int) getField(pos);
+    }
+
+    @Override
+    public long getLong(int pos) {
+        return (long) getField(pos);
+    }
+
+    @Override
+    public float getFloat(int pos) {
+        return (float) getField(pos);
+    }
+
+    @Override
+    public double getDouble(int pos) {
+        return (double) getField(pos);
+    }
+
+    @Override
+    public BinaryString getString(int pos) {
+        return (BinaryString) getField(pos);
+    }
+
+    @Override
+    public Decimal getDecimal(int pos, int precision, int scale) {
+        return (Decimal) getField(pos);
```

Comment:
`scale` is missing here, how about throw exception here? Otherwise, it may return invalid result

Reply:
This is just like `GenericRow`, we trust the decimal returns by `suppliers` has correct precision and scale.